analyzers/src/SonarAnalyzer.Common/Rules/NonStandardCryptographicAlgorithmsShouldNotBeUsedBase.cs

```diff
@@ -0,0 +1,88 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class NonStandardCryptographicAlgorithmsShouldNotBeUsedBase<TSyntaxKind, TTypeDeclarationSyntax> : HotspotDiagnosticAnalyzer
+        where TTypeDeclarationSyntax : SyntaxNode
+        where TSyntaxKind : struct
+    {
+        internal const string DiagnosticId = "S2257";
+
+        protected const string MessageFormat = "Make sure using a non-standard cryptographic algorithm is safe here.";
+
+        private readonly ImmutableArray<KnownType> nonInheritableClassesAndInterfaces = ImmutableArray.Create<KnownType>(
+            new[]
+            {
+                KnownType.System_Security_Cryptography_AsymmetricAlgorithm,
+                KnownType.System_Security_Cryptography_AsymmetricKeyExchangeDeformatter,
+                KnownType.System_Security_Cryptography_AsymmetricKeyExchangeFormatter,
+                KnownType.System_Security_Cryptography_AsymmetricSignatureDeformatter,
+                KnownType.System_Security_Cryptography_AsymmetricSignatureFormatter,
+                KnownType.System_Security_Cryptography_DeriveBytes,
+                KnownType.System_Security_Cryptography_HashAlgorithm,
+                KnownType.System_Security_Cryptography_ICryptoTransform,
+                KnownType.System_Security_Cryptography_SymmetricAlgorithm,
+            });
+
+        protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
+
+        protected abstract TSyntaxKind[] SyntaxKinds { get; }
+
+        protected abstract DiagnosticDescriptor Rule { get; }
+
+        protected abstract Location GetLocation(TTypeDeclarationSyntax typeDeclarationSyntax);
+
+        protected abstract bool DerivesOrImplementsAny(TTypeDeclarationSyntax typeDeclarationSyntax);
+
+        protected abstract INamedTypeSymbol GetDeclaredSymbol(TTypeDeclarationSyntax typeDeclarationSyntax, SemanticModel semanticModel);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected NonStandardCryptographicAlgorithmsShouldNotBeUsedBase(IAnalyzerConfiguration analyzerConfiguration)
+            : base(analyzerConfiguration)
+        {
+        }
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(GeneratedCodeRecognizer, AnalizeDeclaration, SyntaxKinds);
```

Comment:
For security hotspots we need to manually check if the rule is active or not. You can take a look at `ObjectShouldBeInitializedCorrectlyBase` to see how this is handled:
https://github.com/SonarSource/sonar-dotnet/blob/3ab2e10b4effa8760b3ecd4fbb2b3d59f43f1b7f/analyzers/src/SonarAnalyzer.CSharp/Rules/ObjectShouldBeInitializedCorrectlyBase.cs#L49-L52